In [3]:
# Loading necessary libraries

import pandas as pd
import numpy as numpy
from zipfile import ZipFile
from functools import reduce
import matplotlib.pyplot as plt
from scipy import stats
import os

In [4]:
path = r"C:\Users\ncc\Downloads\archive (1).zip"

extracted_file = "data_files"

In [5]:
with ZipFile(path, "r") as f:
    f.extractall("data_files")

In [6]:
data_frames = {}

for file in os.listdir(extracted_file):
    if file.endswith("csv"):
        file_path = os.path.join(extracted_file,file)
        data_frames[file] = pd.read_csv(file_path)

In [7]:
for name, df in data_frames.items():
    print(f"{name}")

olist_customers_dataset.csv
olist_geolocation_dataset.csv
olist_orders_dataset.csv
olist_order_items_dataset.csv
olist_order_payments_dataset.csv
olist_order_reviews_dataset.csv
olist_products_dataset.csv
olist_sellers_dataset.csv
possum.csv
product_category_name_translation.csv


In [8]:
customers = pd.read_csv(r"data_files\olist_customers_dataset.csv")
print(customers.shape)
print(customers.columns)

(99441, 5)
Index(['customer_id', 'customer_unique_id', 'customer_zip_code_prefix',
       'customer_city', 'customer_state'],
      dtype='object')


In [9]:
customers.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [10]:
geolocation = pd.read_csv(r"data_files\olist_geolocation_dataset.csv")
geolocation.head()
print(geolocation.shape)
print(geolocation.columns)

(1000163, 5)
Index(['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng',
       'geolocation_city', 'geolocation_state'],
      dtype='object')


In [11]:
order_items = pd.read_csv(r"data_files\olist_order_items_dataset.csv")
order_items.head()
print(order_items.shape)
print(order_items.columns)

(112650, 7)
Index(['order_id', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value'],
      dtype='object')


In [12]:
order_payment = pd.read_csv(r"data_files\olist_order_payments_dataset.csv")
order_payment.head()
print(order_payment.columns)
print(order_payment.shape)

Index(['order_id', 'payment_sequential', 'payment_type',
       'payment_installments', 'payment_value'],
      dtype='object')
(103886, 5)


In [13]:
order_reviews = pd.read_csv(r"data_files\olist_order_reviews_dataset.csv")
order_reviews.head()
print(order_reviews.columns)
print(order_reviews.shape)

Index(['review_id', 'order_id', 'review_score', 'review_comment_title',
       'review_comment_message', 'review_creation_date',
       'review_answer_timestamp'],
      dtype='object')
(99224, 7)


In [14]:
order_data = pd.read_csv(r"data_files\olist_orders_dataset.csv")
order_data.head()
print(order_data.columns)
print(order_data.shape)

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='object')
(99441, 8)


In [15]:
product = pd.read_csv(r"data_files\olist_products_dataset.csv")
product.head()
print(product.columns)
print(product.shape)

Index(['product_id', 'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm'],
      dtype='object')
(32951, 9)


In [16]:
sellers = pd.read_csv(r"data_files\olist_sellers_dataset.csv")
sellers.head()
print(order_reviews.columns)
print(order_reviews.shape)

Index(['review_id', 'order_id', 'review_score', 'review_comment_title',
       'review_comment_message', 'review_creation_date',
       'review_answer_timestamp'],
      dtype='object')
(99224, 7)


In [17]:
product_category = pd.read_csv(r"data_files\product_category_name_translation.csv")
product_category.head()
print(product_category.columns)
print(product_category.shape)

Index(['product_category_name', 'product_category_name_english'], dtype='object')
(71, 2)


In [18]:
for name, df in data_frames.items():
    print(f"{name}: {list(df.columns)}")

olist_customers_dataset.csv: ['customer_id', 'customer_unique_id', 'customer_zip_code_prefix', 'customer_city', 'customer_state']
olist_geolocation_dataset.csv: ['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng', 'geolocation_city', 'geolocation_state']
olist_orders_dataset.csv: ['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']
olist_order_items_dataset.csv: ['order_id', 'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date', 'price', 'freight_value']
olist_order_payments_dataset.csv: ['order_id', 'payment_sequential', 'payment_type', 'payment_installments', 'payment_value']
olist_order_reviews_dataset.csv: ['review_id', 'order_id', 'review_score', 'review_comment_title', 'review_comment_message', 'review_creation_date', 'review_answer_timestamp']
olist_products_dataset.csv: ['product_id', 'product_category_name', 'pro

In [19]:
# Get sets of columns for each file
columns_per_file = {name: set(df.columns) for name, df in data_frames.items()}

# Intersection (columns that appear in every file)
common_cols = reduce(set.intersection, columns_per_file.values())
print("Common columns across all CSVs:", common_cols)


Common columns across all CSVs: set()


In [20]:

#1️. Customers + Orders
orders_customers = pd.merge(order_data, customers, on="customer_id", how="inner")
orders_customers.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1           2018-08-13 00:00:00  af07308b275d755c9edb36a90c618231   
2           2018-09-04 00:00:00  3a653a41f6f9fc3d2a113cf8398680e8   
3           2017-12-15 00:00:00  7c142cf63193a1473d2e66489a9ae977   
4           2018-02-26 00:00:00  72632f0f9dd73dfee390c9b22eb56dd6   

   customer_zip_code_prefix            customer_city customer_state  
0                      3149                sao paulo             SP  
1                     47813                barreiras             BA  
2                     75265               vianopolis             GO  
3                     59296  sao goncalo do amarante             RN  
4                      9195              santo andre             SP

In [21]:

#2. Orders + Payments
orders_payments = pd.merge(orders_customers, order_payment, on="order_id", how="left")
orders_payments.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
2           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
3           2018-08-13 00:00:00  af07308b275d755c9edb36a90c618231   
4           2018-09-04 00:00:00  3a653a41f6f9fc3d2a113cf8398680e8   

   customer_zip_code_prefix customer_city customer_state  payment_sequential  \
0                      3149     sao paulo             SP                 1.0   
1                      3149     sao paulo             SP                 3.0   
2                      3149     sao paulo             SP                 2.0   
3                     47813     barreiras             BA                 1.0   
4                     75265    vianopolis             GO                 1.0   

  payment_type  payment_installments  payment_value  
0  credit_card                   1.0          18.12  
1      voucher                   1.0           2.00  
2      voucher                   1.0          18.59  
3       boleto                   1.0         141.46  
4  credit_card                   3.0         179.12

In [22]:
#3️. Orders + Reviews
orders_reviews = pd.merge(orders_payments, order_reviews, on="order_id", how="left")
order_reviews.head()

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

In [23]:
#4. Orders + Order Items
orders_items = pd.merge(orders_reviews, order_items, on="order_id", how="left")
orders_items.head()


order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
2           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
3           2018-08-13 00:00:00  af07308b275d755c9edb36a90c618231   
4           2018-09-04 00:00:00  3a653a41f6f9fc3d2a113cf8398680e8   

   customer_zip_code_prefix  ... review_comment_title  \
0                      3149  ...                  NaN   
1                      3149  ...                  NaN   
2                      3149  ...                  NaN   
3                     47813  ...     Muito boa a loja   
4                     75265  ...                  NaN   

                              review_comment_message  review_creation_date  \
0  Não testei o produto ainda, mas ele veio corre...   2017-10-11 00:00:00   
1  Não testei o produto ainda, mas ele veio corre...   2017-10-11 00:00:00   
2  Não testei o produto ainda, mas ele veio corre...   2017-10-11 00:00:00   
3                               Muito bom o produto.   2018-08-08 00:00:00   
4                                                NaN   2018-08-18 00:00:00   

  review_answer_timestamp  order_item_id                        product_id  \
0     2017-10-12 03:43:48            1.0  87285b34884572647811a353c7ac498a   
1     2017-10-12 03:43:48            1.0  87285b34884572647811a353c7ac498a   
2     2017-10-12 03:43:48            1.0  87285b34884572647811a353c7ac498a   
3     2018-08-08 18:37:50            1.0  595fac2a385ac33a80bd5114aec74eb8   
4     2018-08-22 19:07:58            1.0  aa4383b373c6aca5d8797843e5594415   

                          seller_id  shipping_limit_date   price freight_value  
0  3504c0cb71d7fa48d967e0e4c94d59d9  2017-10-06 11:07:15   29.99          8.72  
1  3504c0cb71d7fa48d967e0e4c94d59d9  2017-10-06 11:07:15   29.99          8.72  
2  3504c0cb71d7fa48d967e0e4c94d59d9  2017-10-06 11:07:15   29.99          8.72  
3  289cdb325fb7e7f891c38608bf9e0962  2018-07-30 03:24:27  118.70         22.76  
4  4869f7a5dfa277a7dca6462dcf3b52b2  2018-08-13 08:55:23  159.90         19.22  

[5 rows x 28 columns]

In [24]:

#5️. Order Items + Products
orders_products = pd.merge(orders_items, product, on="product_id", how="left")
orders_products.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
2           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
3           2018-08-13 00:00:00  af07308b275d755c9edb36a90c618231   
4           2018-09-04 00:00:00  3a653a41f6f9fc3d2a113cf8398680e8   

   customer_zip_code_prefix  ...   price freight_value  product_category_name  \
0                      3149  ...   29.99          8.72  utilidades_domesticas   
1                      3149  ...   29.99          8.72  utilidades_domesticas   
2                      3149  ...   29.99          8.72  utilidades_domesticas   
3                     47813  ...  118.70         22.76             perfumaria   
4                     75265  ...  159.90         19.22             automotivo   

  product_name_lenght  product_description_lenght  product_photos_qty  \
0                40.0                       268.0                 4.0   
1                40.0                       268.0                 4.0   
2                40.0                       268.0                 4.0   
3                29.0                       178.0                 1.0   
4                46.0                       232.0                 1.0   

  product_weight_g  product_length_cm product_height_cm product_width_cm  
0            500.0               19.0               8.0             13.0  
1            500.0               19.0               8.0             13.0  
2            500.0               19.0               8.0             13.0  
3            400.0               19.0              13.0             19.0  
4            420.0               24.0              19.0             21.0  

[5 rows x 36 columns]

In [25]:
#6. Products + Product Category Translation
orders_products = pd.merge(
    orders_products, product_category, on="product_category_name", how="left"
)
orders_products.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
2           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
3           2018-08-13 00:00:00  af07308b275d755c9edb36a90c618231   
4           2018-09-04 00:00:00  3a653a41f6f9fc3d2a113cf8398680e8   

   customer_zip_code_prefix  ... freight_value  product_category_name  \
0                      3149  ...          8.72  utilidades_domesticas   
1                      3149  ...          8.72  utilidades_domesticas   
2                      3149  ...          8.72  utilidades_domesticas   
3                     47813  ...         22.76             perfumaria   
4                     75265  ...         19.22             automotivo   

   product_name_lenght product_description_lenght  product_photos_qty  \
0                 40.0                      268.0                 4.0   
1                 40.0                      268.0                 4.0   
2                 40.0                      268.0                 4.0   
3                 29.0                      178.0                 1.0   
4                 46.0                      232.0                 1.0   

   product_weight_g product_length_cm  product_height_cm product_width_cm  \
0             500.0              19.0                8.0             13.0   
1             500.0              19.0                8.0             13.0   
2             500.0              19.0                8.0             13.0   
3             400.0              19.0               13.0             19.0   
4             420.0              24.0               19.0             21.0   

  product_category_name_english  
0                    housewares  
1                    housewares  
2                    housewares  
3                     perfumery  
4                          auto  

[5 rows x 37 columns]

In [26]:

#7️. Order Items + Sellers
orders_sellers = pd.merge(orders_products, sellers, on="seller_id", how="left")
orders_sellers.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
2           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
3           2018-08-13 00:00:00  af07308b275d755c9edb36a90c618231   
4           2018-09-04 00:00:00  3a653a41f6f9fc3d2a113cf8398680e8   

   customer_zip_code_prefix  ... product_description_lenght  \
0                      3149  ...                      268.0   
1                      3149  ...                      268.0   
2                      3149  ...                      268.0   
3                     47813  ...                      178.0   
4                     75265  ...                      232.0   

  product_photos_qty  product_weight_g product_length_cm  product_height_cm  \
0                4.0             500.0              19.0                8.0   
1                4.0             500.0              19.0                8.0   
2                4.0             500.0              19.0                8.0   
3                1.0             400.0              19.0               13.0   
4                1.0             420.0              24.0               19.0   

   product_width_cm product_category_name_english  seller_zip_code_prefix  \
0              13.0                    housewares                  9350.0   
1              13.0                    housewares                  9350.0   
2              13.0                    housewares                  9350.0   
3              19.0                     perfumery                 31570.0   
4              21.0                          auto                 14840.0   

      seller_city seller_state  
0            maua           SP  
1            maua           SP  
2            maua           SP  
3  belo horizonte           SP  
4         guariba           SP  

[5 rows x 40 columns]

In [27]:

#8. Customers + Geolocation (via zip prefix)
#  Note: many geolocations per ZIP, so we take first occurrence per prefix
geo_unique = geolocation.groupby("geolocation_zip_code_prefix").first().reset_index()

final_df = pd.merge(
    orders_sellers,
    geo_unique,
    left_on="customer_zip_code_prefix",
    right_on="geolocation_zip_code_prefix",
    how="left"
)

print(final_df.shape)
final_df.head()

(119143, 45)


order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
2           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
3           2018-08-13 00:00:00  af07308b275d755c9edb36a90c618231   
4           2018-09-04 00:00:00  3a653a41f6f9fc3d2a113cf8398680e8   

   customer_zip_code_prefix  ... product_width_cm  \
0                      3149  ...             13.0   
1                      3149  ...             13.0   
2                      3149  ...             13.0   
3                     47813  ...             19.0   
4                     75265  ...             21.0   

  product_category_name_english  seller_zip_code_prefix     seller_city  \
0                    housewares                  9350.0            maua   
1                    housewares                  9350.0            maua   
2                    housewares                  9350.0            maua   
3                     perfumery                 31570.0  belo horizonte   
4                          auto                 14840.0         guariba   

   seller_state  geolocation_zip_code_prefix geolocation_lat  geolocation_lng  \
0            SP                       3149.0      -23.574809       -46.587471   
1            SP                       3149.0      -23.574809       -46.587471   
2            SP                       3149.0      -23.574809       -46.587471   
3            SP                      47813.0      -12.169860       -44.988369   
4            SP                      75265.0      -16.746337       -48.514624   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        barreiras                BA  
4       vianopolis                GO  

[5 rows x 45 columns]

### **PROBABILITY**


#### 1. Probability of an Event

Concept: Basic probability = favorable outcomes / total outcomes.

Question: What is the probability that a randomly chosen order received a 5-star review?

In [28]:
# Let's use the order_review data to answer this question
orders_reviews.head(2)

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   

   customer_zip_code_prefix  ... payment_sequential payment_type  \
0                      3149  ...                1.0  credit_card   
1                      3149  ...                3.0      voucher   

   payment_installments payment_value                         review_id  \
0                   1.0         18.12  a54f0611adc9ed256b57ede6b6eb5114   
1                   1.0          2.00  a54f0611adc9ed256b57ede6b6eb5114   

   review_score review_comment_title  \
0           4.0                  NaN   
1           4.0                  NaN   

                              review_comment_message review_creation_date  \
0  Não testei o produto ainda, mas ele veio corre...  2017-10-11 00:00:00   
1  Não testei o produto ainda, mas ele veio corre...  2017-10-11 00:00:00   

  review_answer_timestamp  
0     2017-10-12 03:43:48  
1     2017-10-12 03:43:48  

[2 rows x 22 columns]

In [29]:
# Probability of 5-star review
#Out of all orders, what is the probability that a review has a 5-star rating?
#This tells us how likely a customer is to give the maximum satisfaction score.

# p(5star) = number of 5-stars review/total reviews

total_reviews = len(order_reviews)

favourable = (order_reviews["review_score"] == 5).sum()

p_5star = favourable/total_reviews

print(f"probability of 5 stars = {(p_5star *100):.2f}%")



probability of 5 stars = 57.78%


**Interpretation of result**

Approximately 58% of customers give 5 stars. This is the baseline "success" probability of excellent reviews.
_______________________________________________________________
_______________________________________________________________

#### 2. Complementary Probability

Concept: 
P(A complement) = 1 - P(A)

Question: What is the probability that a review is not 5 stars?

In [30]:
# If we know the probability of 5 stars, what's the probability of not getting a 5-star review?
# That's "any rating from 1 to 4"

# We still maintain our review dataset
# Formula style
p_not5 = (order_reviews["review_score"] != 5).sum() / len(order_reviews)
print(f"P(Review ≠ 5 stars) = {round(p_not5 *100, 2)}%")

# Another method to consider
print(f"The complement of P(A) is {round((1-p_5star)*100,2)}%")

P(Review ≠ 5 stars) = 42.22%
The complement of P(A) is 42.22%


**Interpretation**

Approximately 42% of reviews are less than perfect. This highlights that not everyone is fully satisfied with product purchased

#### 3. Joint Probability

Concept: P(A∩B)

Question: What is the probability that an order was paid by credit card and received a 5-star review?

In [31]:
#This gives the likelihood of two events happening together.

# Here we will be using the orders_payments and orders_reviews. And we have merged them already
# lets merge orders_payments and order_reviews
order_reviews.head(2)

orders_payment_reviews = pd.merge(orders_reviews, orders_payments, on="order_id", how="left")
orders_payment_reviews.head()



order_id                     customer_id_x  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
4  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   

  order_status_x order_purchase_timestamp_x  order_approved_at_x  \
0      delivered        2017-10-02 10:56:33  2017-10-02 11:07:15   
1      delivered        2017-10-02 10:56:33  2017-10-02 11:07:15   
2      delivered        2017-10-02 10:56:33  2017-10-02 11:07:15   
3      delivered        2017-10-02 10:56:33  2017-10-02 11:07:15   
4      delivered        2017-10-02 10:56:33  2017-10-02 11:07:15   

  order_delivered_carrier_date_x order_delivered_customer_date_x  \
0            2017-10-04 19:55:00             2017-10-10 21:25:13   
1            2017-10-04 19:55:00             2017-10-10 21:25:13   
2            2017-10-04 19:55:00             2017-10-10 21:25:13   
3            2017-10-04 19:55:00             2017-10-10 21:25:13   
4            2017-10-04 19:55:00             2017-10-10 21:25:13   

  order_estimated_delivery_date_x              customer_unique_id_x  \
0             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
2             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
3             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
4             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   

   customer_zip_code_prefix_x  ... order_delivered_customer_date_y  \
0                        3149  ...             2017-10-10 21:25:13   
1                        3149  ...             2017-10-10 21:25:13   
2                        3149  ...             2017-10-10 21:25:13   
3                        3149  ...             2017-10-10 21:25:13   
4                        3149  ...             2017-10-10 21:25:13   

  order_estimated_delivery_date_y              customer_unique_id_y  \
0             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
2             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
3             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
4             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   

  customer_zip_code_prefix_y  customer_city_y  customer_state_y  \
0                       3149        sao paulo                SP   
1                       3149        sao paulo                SP   
2                       3149        sao paulo                SP   
3                       3149        sao paulo                SP   
4                       3149        sao paulo                SP   

  payment_sequential_y  payment_type_y payment_installments_y payment_value_y  
0                  1.0     credit_card                    1.0           18.12  
1                  3.0         voucher                    1.0            2.00  
2                  2.0         voucher                    1.0           18.59  
3                  1.0     credit_card                    1.0           18.12  
4                  3.0         voucher                    1.0            2.00  

[5 rows x 37 columns]

In [32]:
orders_payment_reviews.columns

Index(['order_id', 'customer_id_x', 'order_status_x',
       'order_purchase_timestamp_x', 'order_approved_at_x',
       'order_delivered_carrier_date_x', 'order_delivered_customer_date_x',
       'order_estimated_delivery_date_x', 'customer_unique_id_x',
       'customer_zip_code_prefix_x', 'customer_city_x', 'customer_state_x',
       'payment_sequential_x', 'payment_type_x', 'payment_installments_x',
       'payment_value_x', 'review_id', 'review_score', 'review_comment_title',
       'review_comment_message', 'review_creation_date',
       'review_answer_timestamp', 'customer_id_y', 'order_status_y',
       'order_purchase_timestamp_y', 'order_approved_at_y',
       'order_delivered_carrier_date_y', 'order_delivered_customer_date_y',
       'order_estimated_delivery_date_y', 'customer_unique_id_y',
       'customer_zip_code_prefix_y', 'customer_city_y', 'customer_state_y',
       'payment_sequential_y', 'payment_type_y', 'payment_installments_y',
       'payment_value_y'],
      dt

In [33]:
total_orders = len(orders_payment_reviews)
favourable_joint = ((orders_payment_reviews["payment_type_y"]=="credit_card") &
                    (orders_payment_reviews["review_score"]==5)).sum()

p_joint = favourable_joint / total_orders

print(f"P(Credit Card ∩ 5 stars) = {round(p_joint *100, 2)}%")


P(Credit Card ∩ 5 stars) = 37.51%


**Interpretation**

Approximately 38% of all orders are both paid by credit card and get a perfect review.

#### 4. Conditional Probability

Concept:

`𝑃(𝐴∣𝐵)=𝑃(𝐴∩𝐵)/𝑃(𝐵)` 
                
`𝑃(𝐴∣𝐵)= Favourable_outcomes / Total_credit_card_outcomes`



Question: What is the probability that a review is 5 stars given that the payment was made by credit card?

In [34]:
"""
If we already know that the payment method was credit card, what is the probability that the review is 5 stars?
This measures satisfaction among credit card users only.
We are solving for independent conditional probability.

To understand better, we can solve further by checking for other means of payments against 5 star reviews

𝑃(5∣credit_card)
P(5∣voucher)
P(5∣boleto)

Each is a separate conditional distribution. They are not mutually exclusive outcomes of the same experiment.
"""

favourable_cond = ((orders_payment_reviews["payment_type_y"] == "credit_card") & (orders_payment_reviews["review_score"] == 5)).sum()

total_credit = (orders_payment_reviews["payment_type_y"] == "credit_card").sum()

p_cond = favourable_cond / total_credit

print(f"P(5 stars | Credit Card) = {round(p_cond * 100, 2)}%")


# For Vouchers
favourable_cond = ((orders_payment_reviews["payment_type_y"]=="voucher") &
                   (orders_payment_reviews["review_score"]==5)).sum()

total_credit = (orders_payment_reviews["payment_type_y"]=="voucher").sum()

p_cond = favourable_cond / total_credit

print(f"P(5 stars | voucher) = {round(p_cond *100, 2)}%")


# For Boleto
favourable_cond = ((orders_payment_reviews["payment_type_y"]=="boleto") &
                   (orders_payment_reviews["review_score"]==5)).sum()

total_credit = (orders_payment_reviews["payment_type_y"]=="boleto").sum()

p_cond = favourable_cond / total_credit
print(f"P(5 stars | boleto) = {round(p_cond *100, 2)}%")

P(5 stars | Credit Card) = 57.39%
P(5 stars | voucher) = 55.71%
P(5 stars | boleto) = 56.95%


**Interpretation**<br><br>
Each one of the payment method is around 55-57%, meaning in every payment method, about half of customers gave 5 stars.

#### 5. Bayes' Rule
Concept: 

P(A|B) = P(B|A).P(A) / P(B)

A = Payment

B = review = 5

P(payment | review=5) = P(review=5 | payment).P(payment) / P(review=5)

Question: Given that an order received a 5-star review, what is the probability that it was paid with each payment method (credit_card, boleto, voucher)?

Among only the 5-star reviews, how are payment methods distributed?

In [ ]:
# Among orders that got a 5-star review, what fraction used each payment method (credit_card, boleto, voucher, etc.)?
# That is, if we look at only 5-star reviews,  how are they split across payment types?

# Total number of 5-star reviews (denominator for all)
total_5star = (orders_payment_reviews["review_score"] == 5).sum()

# ---- P(payment = credit_card | review = 5) ----
favourable_cond = ((orders_payment_reviews["payment_type_y"]=="credit_card") & 
                   (orders_payment_reviews["review_score"]==5)).sum()

p_credit_card = favourable_cond / total_5star
print(f"P(Credit Card | 5 stars) = {round(p_credit_card * 100, 2)}%")

# ---- P(payment = voucher | review = 5) ----
favourable_cond = ((orders_payment_reviews["payment_type_y"]=="voucher") & 
                   (orders_payment_reviews["review_score"]==5)).sum()

p_voucher = favourable_cond / total_5star
print(f"P(Voucher | 5 stars) = {round(p_voucher * 100, 2)}%")

# ---- P(payment = boleto | review = 5) ----
favourable_cond = ((orders_payment_reviews["payment_type_y"]=="boleto") & 
                   (orders_payment_reviews["review_score"]==5)).sum()

p_boleto = favourable_cond / total_5star
print(f"P(Boleto | 5 stars) = {round(p_boleto * 100, 2)}%")

# ---- P(payment = debit card | review = 5) ----
favourable_cond = ((orders_payment_reviews["payment_type_y"]=="debit_card") & 
                   (orders_payment_reviews["review_score"]==5)).sum()

p_debit_card = favourable_cond / total_5star
print(f"P(credit_card | 5 stars) = {round(p_debit_card * 100, 2)}%")

# ---- P(payment = not_defined | review = 5) ----
favourable_cond = ((orders_payment_reviews["payment_type_y"]=="not_defined") & 
                   (orders_payment_reviews["review_score"]==5)).sum()

p_not_defined = favourable_cond / total_5star
print(f"P(not_defined | 5 stars) = {round(p_not_defined * 100, 2)}%")

# The aggregate value for the payment methods
print(f"\nTotal = {round((p_credit_card + p_voucher + p_debit_card + p_boleto + p_not_defined) *100, 2)}%")

P(Credit Card | 5 stars) = 65.74%
P(Voucher | 5 stars) = 16.9%
P(Boleto | 5 stars) = 16.05%
P(debit_card | 5 stars) = 1.31%
P(not_defined | 5 stars) = 0.0%

Total = 100.0%


**Interpretation**<br><br>
1. `P(Credit Card | 5 stars)` = Among all the 5-star reviews, `65.74%` used credit card.

2. `P(Voucher | 5 stars)` = Among all the 5-star reviews, `16.9%` used voucher.

3. `P(Boleto | 5 stars)` = Among all the 5-star reviews, `16.05%` of 5-star reviews uses boleto payment method.

4. `P(Debit Card | 5 stars)` = Among all the 5-star reviews, `1.31%` used debit card payment method.


Unlike P(5 | payment)- conditional probability , these will sum to approximately 100% across all payment methods, because every 5-star order must fall into exactly one payment category.

**I hope the interpretation or difference is clear enough, if not, please, go through it again**

##### Just playing around with the below


In [56]:
# Just playing around with this!!!

# Checking for prob. of getting credit_card
total_payment_type = len(orders_payment_reviews["payment_type_y"])

p_credit_card = (orders_payment_reviews["payment_type_y"] == "credit_card").sum() / total_payment_type

print("Probability of Credit Card: ", round(p_credit_card *100, 2))

# Checking for prob. of getting debit_card
total_payment_type = len(orders_payment_reviews["payment_type_y"])

p_debit_card = (orders_payment_reviews["payment_type_y"] == "debit_card").sum() / total_payment_type

print("Probability of Debit Card: ", round(p_debit_card *100, 2))

# Checking for prob. of getting voucher
total_payment_type = len(orders_payment_reviews["payment_type_y"])

p_voucher = (orders_payment_reviews["payment_type_y"] == "voucher").sum() / total_payment_type

print("Probability of voucher Card: ", round(p_voucher *100, 2))

# Checking for prob. of getting boleto
total_payment_type = len(orders_payment_reviews["payment_type_y"])

p_boleto = (orders_payment_reviews["payment_type_y"] == "boleto").sum() / total_payment_type

print("Probability of boleto Card: ", round(p_boleto *100, 2))

# Checking for prob. of getting not_defined
total_payment_type = len(orders_payment_reviews["payment_type_y"])

p_not_defined = (orders_payment_reviews["payment_type_y"] == "not_defined").sum() / total_payment_type

print("Probability of not_defined Card: ", round(p_not_defined *100, 2))



Probability of Credit Card:  65.37
Probability of Debit Card:  1.24
Probability of voucher Card:  17.31
Probability of boleto Card:  16.08
Probability of not_defined Card:  0.0


In [37]:
orders_payment_reviews["payment_type_y"].value_counts()

payment_type_y
credit_card    80942
voucher        21431
boleto         19910
debit_card      1539
not_defined        3
Name: count, dtype: int64

### **Expectation and Variance**